File: WinForms/CS/HtmlMainDemo/Modules/TreeList/TreeListModule.cs
using System;
using System.ComponentModel;
using DevExpress.DXperience.Demos;
using DevExpress.HTML.Demos.Helpers.Data;
using DevExpress.HTML.Demos.Modules.TreeList;
using DevExpress.Utils.Html;
using DevExpress.XtraEditors.Controls;
using DevExpress.XtraTreeList;
using DevExpress.XtraTreeList.Nodes;

namespace DevExpress.HTML.Demos {
    public partial class TreeListModule : TutorialControlBase {
        EditingComment editingComment;
        BindingList<Comment> dataSource;

        public TreeListModule() {
            InitializeComponent();
            treeListTemplateCodeViewer1.Fill(treeList1);
            dataSource = TreeListData.GenerateCommentsData();
            treeList1.DataSource = dataSource;
            
            var autoHeightEdit = messageEdit as IAutoHeightControlEx;
            autoHeightEdit.AutoHeightEnabled = true;
            autoHeightEdit.HeightChanged += OnMessageHeightChanged;
        }
        protected override int TakeScreenDelay => 300;
        void ExpandCollapseClick(object sender, DxHtmlElementMouseEventArgs e) {
            var node = e.SourceItem as TreeListNode;
            if(node.Expanded)
                node.Collapse(true);
            else
                node.Expand(true);
        }

        void EditClick(object sender, DxHtmlElementMouseEventArgs e) {
            var node = e.SourceItem as TreeListNode;
            StartCommentEditor(node, CommentEditingMode.Editing);
        }

        void ReplyClick(object sender, DxHtmlElementMouseEventArgs e) {
            var node = e.SourceItem as TreeListNode;
            StartCommentEditor(node, CommentEditingMode.Replying);
        }

        void CancelClick(object sender, DxHtmlElementMouseEventArgs e) {
            EndCommentEditor(false);
        }

        void ApplyClick(object sender, DxHtmlElementMouseEventArgs e) {
            EndCommentEditor(true);
        }

        void LikeClick(object sender, DxHtmlElementMouseEventArgs e) {
            var node = e.SourceItem as TreeListNode;
            var comment = treeList1.GetRow(node.Id) as Comment;
            comment.ToggleLike();
        }

        void SendClick(object sender, DxHtmlElementMouseEventArgs e) {
            var comment = TreeListData.CurrentUser.CreateComment(messageEdit.Text);
            dataSource.Add(comment);            
            messageEdit.Clear();
            TreeListData.MakeCommentNodeVisible(comment, treeList1.Nodes);
        }

        void StartCommentEditor(TreeListNode node, CommentEditingMode editingMode) {
            EndCommentEditor(false);
            var comment = treeList1.GetRow(node.Id) as Comment;
            editingComment = new EditingComment(node, comment, editingMode);
            treeList1.RefreshNode(node);
            if(node.NextVisibleNode == null)
                treeList1.MakeNodeVisible(node);
            treeList1.ShowEditor("EditingText");
        }

        void EndCommentEditor(bool applyChanges) {
            if(editingComment != null) {
                treeList1.BeginUpdate();
                var editedComment = editingComment;
                editingComment = null;
                treeList1.RefreshNode(editedComment.Node);
                if(applyChanges)
                    editedComment.Apply();                
                treeList1.EndUpdate();
            }
        }

        void OnMessageHeightChanged(object sender, EventArgs e) {
            var contentSize = typingBox.GetContentSize();
            typingBox.Height = Math.Max(30, contentSize.Height);
        }

        private void treeList1_CustomUnboundColumnData(object sender, TreeListCustomColumnDataEventArgs e) {
            if(e.Column.FieldName == "IsMyComment") {
                Comment comment = e.Row as Comment;
                e.Value = comment.Author == TreeListData.CurrentUser.Name;
            }
            else if(e.Column.FieldName == "HasLikes") {
                Comment comment = e.Row as Comment;
                e.Value = comment.Likes > 0;
            }
            else if(e.Column.FieldName == "EditingText") {
                if(editingComment != null) {
                    if(e.IsGetData)
                        e.Value = editingComment.Text;
                    else {
                        editingComment.Text = Convert.ToString(e.Value);
                    }
                }
            }
        }

        private void treeList1_QueryNodeTemplate(object sender, QueryNodeTemplateEventArgs e) {
            var comment = treeList1.GetRow(e.Node.Id);
            if(editingComment != null && editingComment.Comment == comment) {
                if(editingComment.Mode == CommentEditingMode.Editing)
                    e.Template.Assign(editingCommentTemplate);
                else
                    e.Template.Assign(replyingCommentTemplate);
            }
        }
    }
}

File: WinForms/VB/HtmlMainDemo/Modules/TreeList/TreeListModule.vb
Imports System
Imports System.ComponentModel
Imports DevExpress.DXperience.Demos
Imports DevExpress.HTML.Demos.Helpers.Data
Imports DevExpress.HTML.Demos.Modules.TreeList
Imports DevExpress.Utils.Html
Imports DevExpress.XtraEditors.Controls
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Nodes

Namespace DevExpress.HTML.Demos

    Public Partial Class TreeListModule
        Inherits TutorialControlBase

        Private editingComment As EditingComment

        Private dataSource As BindingList(Of Comment)

        Public Sub New()
            InitializeComponent()
            treeListTemplateCodeViewer1.Fill(treeList1)
            dataSource = TreeListData.GenerateCommentsData()
            treeList1.DataSource = dataSource
            Dim autoHeightEdit = TryCast(messageEdit, IAutoHeightControlEx)
            autoHeightEdit.AutoHeightEnabled = True
            AddHandler autoHeightEdit.HeightChanged, AddressOf OnMessageHeightChanged
        End Sub

        Protected Overrides ReadOnly Property TakeScreenDelay As Integer
            Get
                Return 300
            End Get
        End Property

        Private Sub ExpandCollapseClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            Dim node = TryCast(e.SourceItem, TreeListNode)
            If node.Expanded Then
                node.Collapse(True)
            Else
                node.Expand(True)
            End If
        End Sub

        Private Sub EditClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            Dim node = TryCast(e.SourceItem, TreeListNode)
            StartCommentEditor(node, CommentEditingMode.Editing)
        End Sub

        Private Sub ReplyClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            Dim node = TryCast(e.SourceItem, TreeListNode)
            StartCommentEditor(node, CommentEditingMode.Replying)
        End Sub

        Private Sub CancelClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            EndCommentEditor(False)
        End Sub

        Private Sub ApplyClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            EndCommentEditor(True)
        End Sub

        Private Sub LikeClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            Dim node = TryCast(e.SourceItem, TreeListNode)
            Dim comment = TryCast(treeList1.GetRow(node.Id), Comment)
            comment.ToggleLike()
        End Sub

        Private Sub SendClick(ByVal sender As Object, ByVal e As DxHtmlElementMouseEventArgs)
            Dim comment = TreeListData.CurrentUser.CreateComment(messageEdit.Text)
            dataSource.Add(comment)
            messageEdit.Clear()
            Call TreeListData.MakeCommentNodeVisible(comment, treeList1.Nodes)
        End Sub

        Private Sub StartCommentEditor(ByVal node As TreeListNode, ByVal editingMode As CommentEditingMode)
            EndCommentEditor(False)
            Dim comment = TryCast(treeList1.GetRow(node.Id), Comment)
            editingComment = New EditingComment(node, comment, editingMode)
            treeList1.RefreshNode(node)
            If node.NextVisibleNode Is Nothing Then treeList1.MakeNodeVisible(node)
            treeList1.ShowEditor("EditingText")
        End Sub

        Private Sub EndCommentEditor(ByVal applyChanges As Boolean)
            If editingComment IsNot Nothing Then
                treeList1.BeginUpdate()
                Dim editedComment = editingComment
                editingComment = Nothing
                treeList1.RefreshNode(editedComment.Node)
                If applyChanges Then editedComment.Apply()
                treeList1.EndUpdate()
            End If
        End Sub

        Private Sub OnMessageHeightChanged(ByVal sender As Object, ByVal e As EventArgs)
            Dim contentSize = typingBox.GetContentSize()
            typingBox.Height = Math.Max(30, contentSize.Height)
        End Sub

        Private Sub treeList1_CustomUnboundColumnData(ByVal sender As Object, ByVal e As TreeListCustomColumnDataEventArgs)
            If Equals(e.Column.FieldName, "IsMyComment") Then
                Dim comment As Comment = TryCast(e.Row, Comment)
                e.Value = Equals(comment.Author, TreeListData.CurrentUser.Name)
            ElseIf Equals(e.Column.FieldName, "HasLikes") Then
                Dim comment As Comment = TryCast(e.Row, Comment)
                e.Value = comment.Likes > 0
            ElseIf Equals(e.Column.FieldName, "EditingText") Then
                If editingComment IsNot Nothing Then
                    If e.IsGetData Then
                        e.Value = editingComment.Text
                    Else
                        editingComment.Text = Convert.ToString(e.Value)
                    End If
                End If
            End If
        End Sub

        Private Sub treeList1_QueryNodeTemplate(ByVal sender As Object, ByVal e As QueryNodeTemplateEventArgs)
            Dim comment = treeList1.GetRow(e.Node.Id)
            If editingComment IsNot Nothing AndAlso editingComment.Comment Is comment Then
                If editingComment.Mode = CommentEditingMode.Editing Then
                    e.Template.Assign(editingCommentTemplate)
                Else
                    e.Template.Assign(replyingCommentTemplate)
                End If
            End If
        End Sub
    End Class
End Namespace
